stdlib/discreteDistributions.hk

```diff
@@ -0,0 +1,111 @@
+# Hakaru Standard Library
+# Discrete Distributions: 
+#	definitions for distributions with discrete outcomes
+# author: Mahmoud Khattab
+
+###########################
+# Combinatorial Functions #
+###########################
+
+# Definitions for basic Combinatorial Functions used in some functions
+
+# Factorial n! = 1*2*...*(n-1)*n
+def factorial(n nat):
+	product i from 1 to n+1: i
+
+# Choose function: n C k
+# If return w/ real2nat, there are rounding errors
+
+	# OVERFLOW ERRORS with Choose()
+	# Overflow error when n is too high. 
+	#	E.g choose(30,10) returns 3 and 
+	# 	choose(30,15) returns this error:
+
+	#		hakaru: unsafeNat: negative input
+	# 		CallStack (from HasCallStack):
+  	# 		error, called at haskell/Data/Number/Nat.hs:71:19 in hakaru-0.6.0-AbnzUW5EnqjBuQmLn3LJbU:Data.Number.Nat
+
+def choose(n nat, k nat):
+	# TODO error check k <= n
+	(product i from k+1 to n+1: i)/(product i from 1 to int2nat(n-k+1): i)
```

Comment:
Is this a better definition for choose?

Reply:
It is still computationally quite bad.  Best is likely to assume it as a primitive (i.e. that `binomial` takes 2 `nat` and returns a `nat`). And implement that primitive in Haskell, by using the implementation from `Math.Combinatorics.Exact.Binomial` (from the exact-combinatorics package).  See https://hackage.haskell.org/package/exact-combinatorics-0.2.0.8/docs/Math-Combinatorics-Exact-Binomial.html if you are curious.

Reply:
First let me note that what is mathematically the same thing may be best defined differently depending on how the program is going to be interpreted -- by running and sampling, by simplification and human reading, or in some other way. For simplification and human reading, this definition of `choose` in terms of `product` is fine. For running and sampling, if it's ok to have some rounding errors for large inputs, it's better to use `betaFunc`. According to https://en.wikipedia.org/wiki/Beta_function, <img src="https://wikimedia.org/api/rest_v1/media/math/render/svg/05a383ddf708634b9124b0fcf1928c3e389faa75">

Reply:
Which makes me think that we might be better off implementing `choose` as a primitive!

Reply:
It's hard to tell without knowing how it would be used.